### ISSUE TITLE
Dynamic creation of fields on an existing message

### ISSUE THREAD

**Issue by AndrewHogg** [jhump/protoreflect#58]
I've been experimenting with adding fields to existing message descriptors on the fly, there seems no way in which to access the functionality without creating a shim to export the createFieldDescriptor function.

With that exposed, I can take a parent message descriptor + file descriptor and add a new field

alias := descriptor.FieldDescriptorProto_TYPE_DOUBLE
fdp := descriptor.FieldDescriptorProto{
			Name:   proto.String("examplename),
			Number: proto.Int(index),
			Label:  descriptor.FieldDescriptorProto_LABEL_OPTIONAL.Enum(),
			Type:  &alias,
		}
return desc.CreateFieldDescriptor(fileDescriptor, parentDescriptor, "", &fdp)

This then gives me a FieldDescriptor I can further use and places it onto the parent message - index here has to be a suitable position e.g. not clashing with existing fields.

Is the functionality accessible in another form, or should this function be exported?

**Comment by jhump** [comment]
That sounds like a very custom need. I'm leaning strongly away from exporting these methods in this library.

What is the actual use case? If you are dynamically adding things to the message, what are you then doing with the message? It seems unlikely that any other consumer of the message would have those same fields, rendering them inaccessible/unusable. It also seems kind of scary to programmatically generate fields that way since you have no way of knowing how to assign tag numbers that won't conflict with actual fields added to the proto source in the future.

My thoughts on how that should better be handled:

1. Define an extension range on the message that is just for these dynamically added fields.
2. Create the field descriptors by simply generating a small file descriptor.
    * The file would need a unique name.
    * The file would include the original file descriptor (that defines the message) as a dependency.
    * The file's only contents would be extension fields that extend that message, and use tags in the above-described range.
3. You can then use `desc.CreateFileDescriptor(newFieldsFile, messageDesc.GetFile())` to create the file descriptor and retrieve the resulting field descriptors from that.

Depending on the actual type of the field, it could get more complicated -- like needing more dependencies to properly link the field's type if it is a message or enum that is not defined in the same file as the extended message.

If you didn't want to use extensions, you could still use the same basic approach, except that instead of defining extensions in this fabricated file descriptor, you re-define the message -- just with different fields/tags. As long as the file has the same package and uses the same name for the message, the resulting field descriptor is still usable with a dynamic message.

If you want to see some examples of fabricating file descriptors like this, there are a couple in this repo:

1. There are tests that fabricate messages this way, for testing `dynamic.AnyResolver`: https://github.com/jhump/protoreflect/blob/master/dynamic/json_test.go#L57
2. There is also significantly more complicated logic (not for the faint of heart!) that transforms [`google.protobuf.Type`](https://github.com/google/protobuf/blob/master/src/google/protobuf/type.proto#L47) messages into a message descriptors by fabricating file descriptors: https://github.com/jhump/protoreflect/blob/master/dynamic/msgregistry/ptype_resolver.go#L266

**Comment by AndrewHogg** [comment]
Use case is around serializing a set of *sql.Rows - a .Net equivalent is  https://github.com/rdingwall/protobuf-net-data - which generates message structures on the fly without the need for .proto files. That will serialize a dataset into a semi-structured (but not fixed) protobuf message.

The structure it uses is a structured part for the column definitions (a schema definition structure forming part of the message) followed by the row data (the variable part) .

**Comment by jhump** [comment]
Interesting. I really don't like the idea of exporting the existing function -- that breaks encapsulation pretty badly, has potentially non-obvious side effects, and probably isn't even sufficient to do what you are wanting to do. I think this functionality belongs in a different package, a "descriptor builder" package, that allows for programmatic construction of descriptors (with builder-pattern APIs).

Do you have any example code where I could see more specifically how you plan on using this? Do you need the actual descriptors to be mutable? Or could you simply re-construct a new descriptor every time you encounter a new field (effectively copy-on-write)? In a perfect world, you could build the entire message definition once, up-front. But, even if you could not, a dynamic message can be built from one version of a message descriptor and then mutated using fields from a different version (as long as the two versions are compatible -- e.g. later version doesn't try to change the meaning/interpretation of fields that were present in older version).

**Comment by AndrewHogg** [comment]
A separate package sounds suitable, accessing the private method via a shim was less than ideal, but seems to get the job done and code working for now (validation of output pending). The descriptors do not need to mutate once created - they only last the duration of the conversion of the rows to protobuf for that set of sql.Rows. The next set of rows passed in has to construct them all again - even running the same query twice you could not re-use them since it could produce different results on the basis someone could of changed the stored procedure on the DB) .

You can see the message structure that the .net library uses : https://github.com/rdingwall/protobuf-net-data/blob/master/src/ProtoBuf.Data/ProtoDataWriter.cs
// This is the underlying protocol buffers structure we use:
// <1 StartGroup> each DataTable
// <SubItem>
//     <2 StartGroup> each DataColumn
//     <SubItem>
//         <1 String> Column Name
//         <2 Variant> Column ProtoDataType (enum casted to int)
//     </SubItem>
//     <3 StartGroup> each DataRow
//     <SubItem>
//         <(# Column Index) (corresponding type)> Field Value
//     </SubItem>
// </SubItem>
The message contains a column definitions part, that library made the protobuf self describing - the static part of the message provides sufficient information to decode the variable part (which is rather nice). That eliminates any concerns around later versions, since there is no such concept within it, every message has the ability to be different - the actual proto / descriptors are not retained beyond a single serialization of a row set.

I'm looking at this purely as a mechanism for efficient serialization of sql.Rows - in a compatible format to the library.

I've got to validate the protobuf that is being output, but once that is done I should be able to show an example conversion.

**Comment by AndrewHogg** [comment]
I've managed to get to the point where I can confirm the protobuf that is output when I've created the message structure on the fly as above is able to be decoded by the .net library.

**Comment by jhump** [comment]
I've started writing a `desc/builder` package.

It will be much simpler than manually creating full `FileDescriptoroProto` messages, and makes cases like yours -- taking a `*desc.MessageDescriptor` and mutating it -- really simple. And if you do want to fully create file descriptors programmatically, from scratch, it has affordances that make that easier than just building the proto messages from scratch (including a little bit of data validation to make sure you're building a valid thing).

The API will include `New*Builder` factory methods for each kind of descriptor, for creating things from scratch, and `BuilderFrom*` methods, for creating builders that mirror an existing descriptor. After mutating the builders (via exported fields and/or setter methods), you then call a `Build()` method to turn them (back) into full-fledged descriptors.

**Comment by jhump** [comment]
@AndrewHogg, if you pull down the repo and checkout the `builder-package` branch, maybe take it for a test drive and let me know what you think of the API etc.

Note that it's not yet finished, so you may very well find bugs. I've got the branch in a PR, #59, so if you do fine bugs please drop a comment on that PR to let me know what you've encountered.

**Comment by AndrewHogg** [comment]
will do

**Comment by AndrewHogg** [comment]
Hi, I've start constructing the message using the builder package, and it successfully outputs protobuf, although the it can not be consumed into the .net library - looking at the byte level output and code, I believe it is down the message vs group - the output starts 0B (11, type message) but the next byte on the output that works is 13 - this would represent repeated group dataTable = 1  - in the builder output it has a second byte of 0B again, which I believe indicates its stored a nested message. 

There is no FieldTypeGroup method so I quickly added one locally:
func FieldTypeGroup(mb *MessageBuilder) *FieldType {
	return &FieldType{
		fieldType:    dpb.FieldDescriptorProto_TYPE_GROUP,
		localMsgType: mb,
	}
}

but it does not alter the output - I'm going to try next to load a static proto file which have the groups defined, and then use the BuilderFromField for the variable part to get to the builder, add the fields then build.

**Comment by jhump** [comment]
Groups are deprecated. Aside from an encoded type of "group" vs "message" on the wire, the following are equivalent:

```proto
message MessageWithGroup {
  optional group GroupField = 1 {
    optional string name = 1;
    optional uint64 id = 2;
  }
}

message MessageWithoutGroup {
  optional GroupField groupField = 1;
  message GroupField {
    optional string name = 1;
    optional uint64 id = 2;
  }
}
```

I didn't add support in the builder for groups because they are a bit tricky to get right since the referenced message is meant to be referred to *only* by the group field. But I could probably do something there similar to how map entry messages are supported. I didn't really intend to support them, though, since they are deprecated.

Could you just not use groups?

**Comment by AndrewHogg** [comment]
The .net library and downstream is still using groups and expects them on the wire - otherwise yes, I would of ditched the groups immediately.

If I use a preformed message for the fixed part of the structure and then using builder.BuilderFromMessage(groupDescriptor) to add in the unstructured part - it results in protbuf that is very similar, the group structure is in place, but there are some byte level differences in the unstructured part that I'm debugging.

**Comment by jhump** [comment]
Could you put together a small repro case? If you set the field's type to group, I would expect it to work. It is possible it's related to support for group encoding in dynamic message, not in the builders.

**Comment by AndrewHogg** [comment]
It's going to be hard to get it down to a small test case - at present I think the problem on the unstructured part is again group related, date times are sent as a group - I'm creating it on the fly and its encoding it as a message.

**Comment by jhump** [comment]
> It's going to be hard to get it down to a small test case

Sounds like a good debugger will be your best ally then. Sorry. Let me know what you find. Patches welcome!

**Comment by AndrewHogg** [comment]
Got it, I still had a builder.FieldTypeMessage(dateMessage) in that piece of code, when I changed it to call the FieldTypeGroup I added above, the protobuf is identical at a byte level

**Comment by AndrewHogg** [comment]
So modifying an existing message using the builders works and works well (severe kudos to you there, that's complex code) - the only issue I have so far is I can't create the whole message on the fly but need a starting point.

**Comment by jhump** [comment]
Okay, so maybe that's a small enough repro case: trying to create a message on the fly that has a group. I'll see if I can repro the encoding you see with that. But I won't get a chance to play with it until this evening at the earliest.

**Comment by jhump** [comment]
Oh, wait. I missed your previous comment.

So the only issue is that you need a way to construct groups programmatically. That's not so bad.

Thanks for experimenting and providing feedback! I've been cleaning up the implementation a bit. It still needs more doc and tests. I can probably do something with the group this evening. I expect I would just do what you already did: add methods for group field types.

But the extra bit I think is needed may be to link the message builder as a child of the field, so that it can get inlined properly, as a nested message on the field's message type (which is kind of how the map entry messages are done).

**Comment by AndrewHogg** [comment]
Yes, its just if I create a whole message from scratch the groups are not serializing quite correctly, if I start with a proto.file which contains groups like this:
message Dataset {
	repeated group dataTable = 1 {
		repeated group columnDefinition = 2 {
			required string name  = 1;	
			required int32 type = 2;
		}
		repeated group dataRow = 3 {
		}
	}
}

and then use the builder.BuilderFromMessage on the dataRow group, I can extend it and add the fields. This serializes correctly. The only caveat is that I had to add the func FieldTypeGroup(mb *MessageBuilder) *FieldType { method.

The issue is localized down to creating messages with groups in them, the serialization starts 0B 0B, (and contains lots of further message constructs and not group ones)

**Comment by jhump** [comment]
@AndrewHogg, I pushed changes to that `builder-package` branch. Over the past few evenings, I've made quite a lot of changes, mostly in tweaking the API, making the implementation a little more DRY, and adding some doc. It's still not quite done, but I think it's really close.

The latest incarnation includes support for groups. However, there is no `FieldTypeGroup` function, to create a `*FieldType` for a group. Instead, to make it easier to enforce how groups are actually modeled in the language, you create the `*MessageBuilder` and then turn it into a group field using `NewGroupField(*MessageBuilder) *FieldBuilder`.

Like the proto language, the message name that the group is built from must start with a capital letter, and the corresponding field will be the lower-cased form of that name. I haven't tested all of the new logic much, but there is a very basic test that at least verifies that you can use the new `NewGroupField` function, add the result to a `MessageBuilder`, and that it spits out a correct descriptor. (Other manipulations of the builders, including renaming/renumber the group message or field, etc, may not work, since I did add some new logic around those actions.)

**Comment by AndrewHogg** [comment]
Trying out the new code - I can create the groups on the fly, but when coming to try reference in terms of dynamic.NewMessage(myMessage) to create a message to assign as a value to the group field, the FQN of the message descriptor that dynamic.NewMessage gives me is just xxx where as the FQN of the group is zzz.yyy.xxx (since it was used to create a group field and added as a field to the message)

The underlying message FQN being just the name and not including the parents then causes it to fail the check during a SetField. Trying to figure out how I can create a new message that is correct for the checks SetField performs.

**Comment by jhump** [comment]
@AndrewHogg, can you post a code snippet of how you are building the descriptor that you pass to `dynamic.NewMessage`?

**Comment by jhump** [comment]
It sounds like you might be doing something like using a message descriptor on its own (or created independently, with fully-qualified name "xxx"). And from that creating another message that refers to the earlier one as the group type (so that the field ends up referring to a message with fully-qualified name "zzz.yyy.xxx"). But that's not valid. You'd need to use the same message descriptor for both:

```go
// xxxMsgDesc is the message that defines the group
xxx, err := builder.FromMessage(xxxMsgDesc)
if err != nil { panic(err) }
yyy := builder.NewMessage("yyy").
    AddField(NewGroupField(xxx))
builder.NewFile("").
    SetPackage("zzz").
    AddMessage(yyy)
// yyyMsgDesc is the message that has a group field
yyyMsgDesc, err := yyy.Build()
if err != nil { panic(err) }

outerMsg := dynamic.NewMessage(yyyMsgDesc)
field := outerMsg.FindFieldByName("xxx")
innerMsg := dynamic.NewMessage(xxxMsgDesc)

// Uh oh! The value refers to name "xxx" but the
// outer wants a message of type "zzz.yyy.Xxx"
outerMsg.SetField(field, innerMsg)
```

To fix, make sure the dynamic message you are creating is of the right type:

```go
outerMsg := dynamic.NewMessage(yyyMsgDesc)
field := outerMsg.FindFieldByName("xxx")
// use field.GetMessageType(), *not* xxxMsgDesc!
innerMsg := dynamic.NewMessage(field.GetMessageType())

// Now it works!
outerMsg.SetField(field, innerMsg)
```

Does this make sense?

The reason groups work this way is to mimic how the actual proto language enforces them: groups are nested messages that are siblings of the group field (with the same name except group name must start with capital letter, corresponding field name is lower-case).

If you *really* need to have a reference to a message that does *not* become an embedded message, you could try the following:

```go
xxx, err := builder.FromMessage(xxxMsgDesc)
if err != nil { panic(err) }
yyy := builder.NewMessage("yyy").
    AddField(NewGroupField(xxx))
// this is the magic to extract the group message type out as a top-level
// message, instead of nested in the same message that owns the field
builder.Unlink(xxx)
```

**Comment by jhump** [comment]
Should be addressed as of #59.

If you come across bugs, please file different issues in this project.

**Comment by AndrewHogg** [comment]
Can confirm that it works - I was not using the find field by name
